hal/rfcs/0000-spi-overhaul.md

```diff
@@ -0,0 +1,246 @@
+  SPI_BIT_ORDERING_MSB_FIRST,
+  SPI_BIT_ORDERING_LSB_FIRST,
+} spi_bit_ordering_t;
+
+typedef void (*spi_async_handler_f)(spi_t *obj, void *ctx, spi_async_event_t event);
+
+SPIName spi_get_module(PINName MISO, PINName MOSI, PINName MCLK);
+void spi_get_capabilities(SPIName name, PINName SS, spi_capabilities_t *cap);
+
+void spi_init(spi_t *obj, bool is_slave, PINName MISO, PINName MOSI, PINName MCLK, PINName SS);
+void spi_format(spi_t *obj, uint8_t bits, spi_mode_t mode, spi_bit_ordering_t bit_ordering);
+uint32_t spi_frequency(spi_t *obj, uint32_t hz);
+void spi_transfer(spi_t *obj, const void *tx, uint32_t tx_len, void *rx, uint32_t rx_len, const void *fill_symbol);
+bool spi_transfer_async(spi_t *obj, const void *tx, uint32_t tx_len, void *rx, uint32_t rx_len, const void *fill_symbol, spi_async_handler_f handler, void *ctx, DMAUsage hint);
+void spi_transfer_async_abort(spi_t *obj);
+void spi_free(spi_t *obj);
+```
+
+## Behaviours
+### Defined Behaviours
+
+- `spi_get_module()` returns the `SPIName` unique identifier to the peripheral associated to this SPI channel.
+- `spi_get_capabilities()` fills the given `spi_capabilities_t` instance
+- `spi_get_capabilities()` should consider the `SS` pin when evaluation the `support_slave_mode` capability.  
+  If the given `SS` pin cannot be managed by hardware in slave mode, `support_slave_mode` should be false.
+- At least a symbol width of 8bit must be supported.
+- The supported frequency range must include the range [0.2..2] MHz.
+- The shortest part of the duty cycle must not be shorter than 50% of the expected period.
+- `spi_init()` initializes the pins leaving the configuration registers unchanged.
+- `spi_init()` if `is_slave` is false:
+    - if `SS` is `NC` the hal implementation ignores this pin.
+    - if `SS` is not `NC` then the hal implementation owns the pin and its management.
+- When managed by the hal implementation, `SS` is always considered active low.
+- If `miso` (exclusive) or `mosi` is missing in any function that expects pins, the bus is assumed to be half-duplex.
+- `spi_free()` resets the pins to their default state.
+- `spi_free()` disables the peripheral clock.
+- `spi_format()` sets :
+  - the number of bits per symbol
+  - the mode :
+    0. Clock idle state is *low*, data are sampled when the clock becomes *active* (polarity = 0, phase = 0)
+    1. Clock idle state is *low*, data are sampled when the clock becomes *inactive* (polarity = 0, phase = 1)
+    2. Clock idle state is *high*, data are sampled when the clock becomes *active* (polarity = 1, phase = 0)
+    3. Clock idle state is *high*, data are sampled when the clock becomes *inactive* (polarity = 1, phase = 1)
+  - the bit ordering (lsb/msb first).
+- `spi_format()` updates the configuration of the peripheral except the baud rate generator.
+- `spi_frequency()` sets the frequency to use during the transfer.
+- `spi_frequency()` returns the actual frequency that will be used.
```

Comment:
For testing purposes it would be good to define how actual frequency can differ from the requested one.

For example in case of K64F when we request 1 MHz, the real freq will be 937500 Hz.
This gives 6,25% deviation, so maybe 10% tolerance can be used?

Reply:
I don't know if we can set any kind of tolerance here cf https://github.com/ARMmbed/mbed-os/issues/5041#issuecomment-337020084
We added here the ability for the user to check the actual frequency. It is then up to the user to decide if the actual frequency if acceptable or not.

Reply:
At the minute it will give you as fast as it can without exceeding the specified speed. There's no lower bound on how slow it can go. Users can read back and assert if they think it's going to be too slow.

In practice I've found it's not the nominal frequency that's the problem often so much as the set-up time per transfer. That can be massive, and really hurts radio drivers who deal in small packets.

Mind you, I guess you can reason that the speed must be `>= min(maximum hardware speed, requested speed / 2)` - worst case is when you've requested something just a bit slower than maximum and it has to divide by 2.